[quant][fx][graphmode] Run symbolic_trace in quantization

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45920 [quant][fx][graphmode][api] Change API for custom module
* **#45919 [quant][fx][graphmode] Run symbolic_trace in quantization**

Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145857](https://our.internmc.facebook.com/intern/diff/D24145857)

jerryzh168 committed:
[quant][fx][graphmode] Run symbolic_trace in quantization

Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Run symbolic_trace in quantization"


Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145857](https://our.internmc.facebook.com/intern/diff/D24145857)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Run symbolic_trace in quantization"


Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145857](https://our.internmc.facebook.com/intern/diff/D24145857)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/quantize_fx.py:
```diff
@@ -20,23 +23,36 @@ def fuse_fx(graph_module, inplace=False):
-    graph_module = fuse_fx(graph_module, inplace)
+    # symbolically trace the model
+    standalone_modules = qconfig_dict.get('standalone_module_name', [])
+    class CustomTracer(Tracer):
```
Can we move the definition of this class out of the function and instantiate it with standalone_modules as an argument?
```python
class CustomTracer(Tracer):
   def __init__(self, standalone_modules):
      super(CustomTracer,self).__init__()
      self.standalone_modules = standalone_modules

   def is_leaf_module(self, m, module_qualified_name):
            return (m.__module__.startswith('torch.nn') and \
            not isinstance(m, torch.nn.Sequential)) or \
            module_qualified_name in self.standalone_modules

jerryzh168 replied:
sure

raghuramank100 commented on torch/quantization/quantize_fx.py:
````diff
@@ -51,14 +67,32 @@ def _prepare_standalone_module_fx(graph_module, qconfig_dict, inplace=False):
+    m = fuse_fx(m)
+    ```
+    """
+    assert not model.training, 'fuse_fx only works on models in eval mode'
````
Where is fusion done for QAT?

jerryzh168 replied:
this is the top level fusion function. we have another function `_fuse_fx` which does not check this flag. since fusion for QAT only makes sense for quantization so i didn't expose that

raghuramank100 reviewed (changes requested):
Some suggestions, thanks!

jerryzh168 committed:
Update on "[quant][fx][graphmode] Run symbolic_trace in quantization"


Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145857](https://our.internmc.facebook.com/intern/diff/D24145857)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Run symbolic_trace in quantization"


Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145857](https://our.internmc.facebook.com/intern/diff/D24145857)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Run symbolic_trace in quantization"


Summary:
As discussed with JIT team, we'll run symbolic trace in quantization functions
prepare_fx now takes orginal pytorch model (torch.nn.Module) instead of `GraphModule` as input

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24145857](https://our.internmc.facebook.com/intern/diff/D24145857)

[ghstack-poisoned]
